### PR TITLE
Docker all the things

### PR DESCRIPTION
**Pull request by kstrauser**
This adds support for building a Docker image. It also supports docker-compose, which makes building and running Monica potentially as simple as:

```shell
$ docker-compose build
$ docker-compose up
```

I tried to make the Dockerfile as straightforward as possible, but it has a couple of unusual bits:

- Most Dockerfiles try to combine all the apt-get stuff onto a single line with a final command (on the same line) to remove the apt files (see https://www.dockerfile.in for an example). That makes for nice, compact image but it's a pain in the neck while developing.

- Each line in a Dockerfile creates a new hashed disk image, and if you change one line in the file, everything after that line has to be rebuilt. To make build times less painful, I created a chunk of lines to run `npm install` before doing anything else with Monica's source files. Since those files might not change as often as the PHP and CSS files you might be editing locally, it's nice to cache that long process so we can skip it during iterative testing. (I tried to do the same for `composer install` but that turned into a rabbit hole and I temporarily gave up on it.)

### CODE REVIEW

**Comment by bensbox** [comment]
Nice work :) I didn't run it yet, but as a quick suggestion i noticed, that you use alot of RUN commands.
In case docker is still doing that, every time you run a command it will create another read-only-layer, which will increase the overall size of the image.
You could combine commands like
```
RUN apt-get update
RUN apt-get install -y git netcat
RUN apt-get install -y composer npm php7.0 php7.0-gd php7.0-intl php7.0-mbstring php7.0-mysql php7.0-xml php7.0-zip
RUN apt-get install -y apache2 libapache2-mod-php7.0
RUN apt-get install -y telnet vim-tiny
```
to
```
RUN apt-get update && apt-get install -y git netcat composer npm php7.0 php7.0-gd php7.0-intl php7.0-mbstring php7.0-mysql php7.0-xml php7.0-zip apache2 libapache2-mod-php7.0 telnet vim-tiny
```
and reduce the image by 4 layers. You could add "&& apt-get clean" to get rid of the deb packages in the apt-cache as well, before the layer becomes read-only.

Another suggestion would be to use alpine linux as base image. It is far smaller than ubuntu (even though i love ubuntu ;) )

**Comment by kstrauser** [comment]
Thanks! I talked a little about my reasoning for having separate RUN lines in the PR description. TL;DR combining all that stuff into a single line before it's pretty certain that it won't be changing again is a PITA. Adding a new package blows the Docker cache for that single RUN line and requires doing the whole thing again from scratch. I think we definitely _should_ combine them all (probably soon!), but for a first draft I felt like this would make iteration a little quicker.

As to Alpine: I'm totally on board with it, except that it doesn't package composer. There _is_ [an install script](https://getcomposer.org/doc/faqs/how-to-install-composer-programmatically.md) that we could use to smooth that over, so that's not the end of the world. I just wanted to minimize the amount of extra-distro work we'd have to manage on the first try.

**Comment by bensbox** [comment]
Absolutely reasonable as a first version - didn't think about that. Just need to make sure, that the prototype is not becoming the final version :D

**Comment by Kovah** [comment]
Any reason why you choose Apache over Nginx? Also, using the Ubuntu base image takes a lot of unnecessary space.

We use the Docker config for Nginx with PHP7 from [ttaranto](https://github.com/ttaranto/docker-nginx-php7) which works perfectly and is really fast while also maintaining a small container size. 

Maybe you could take a look at it.

**Comment by kstrauser** [comment]
@bensbox I made an Alpine version at https://github.com/kstrauser/monica/tree/feature/alpine-docker. It's still not absolutely compact (for the same reasons as before), but it significantly moreso and weighs in at about 455MB instead of 874MB for Ubuntu. That was easier than I'd feared!

@Kovah I'd already mentioned why I initially used Ubuntu over Alpine: it was easier as a first pass because all requirements were available pre-packaged for it. Even though I love Nginx, I picked Apache for the same reason. Ubuntu (and Alpine) package mod_php for it, and that's a bit easier to get running under Docker than FastCGI or PHP-FPM would be.

The first goal of this PR was to make something simple that non-experts could start playing with quickly. That _absolutely_ doesn't meant I wouldn't want to make a better version later though. I didn't have to change a single line of existing Monica code or config, so I think we could swap out Apache for Nginx later without anyone else even noticing.

**Comment by kstrauser** [comment]
I merged my Alpine branch in. Since it's fully working now, might as well launch with it.

**Comment by kstrauser** [comment]
I created a new monicahq organization on Docker Hub. @djaiss, I'll turn that over to you (or whoever you designate to manage it) whenever you're ready.

With these changes, a new user can do this to be up and running:

- `cp .env.example .env` and edit it
- `docker-compose pull`
- `docker-compose up`

I think that's about as simple as I can make it, short of someone writing a pretty config file generator.

**Comment by djaiss** [comment]
I'm mind blowed by this. Thanks a lot.

Can you provide a link to Docker Hub in this PR (I assume it's not something sensible anyway)?

Is this branch ready to be merged? I haven't tried it yet and won't have time to try it until a few days, considering the amount of stuff I need to do, but Docker is really important for the adoption, so if you assume it's ready, I'll trust you on that (since it doesn't change one line of the app itself, which is awesome).

File: readme.md [hunk 1]
```diff
@@ -64,12 +64,82 @@ money out of it yourself.
...
+## Running with Docker
+
+You can use [Docker](https://www.docker.com) and
+[docker-compose](https://docs.docker.com/compose/) to pull or build
+and run a Monica image, complete with a self-contained MySQL database.
+This has the nice properties that you don't have to install lots of
+software directly onto your system, and you can be up and running
+quickly with a known working environment.
+
+Before you start, run `cp .env.example .env` and open `.env` in an
```
Comment (djaiss) [inline readme.md]: I feel there is a step missing.

Do we need to git clone the folder? Do we need to run a special docker CLI instruction before?

Comment (moosilauke18) [inline readme.md]: To run locally, you would clone then run the following. However, you should also be able to just pull `monicahq/monicahq` and run it basically like `docker run -e DB_HOST=localhost...<other env> -p 80:80 monicahq/monicahq -d`.

**Comment by moosilauke18** [comment]
This works wonderfully for me without docker compose and using a non-dockerized mysql. Just pull from docker hub and put in ENV into run command. Great job!

File: readme.md (hunk 1 above)
Comment (kstrauser) [inline readme.md]: I added some clarification in a82b650.

File: docker/test-server.sh [hunk 2]
```diff
@@ -4,4 +4,5 @@ ARTISAN=/var/www/monica/artisan
 
 php ${ARTISAN} migrate --force
 php ${ARTISAN} storage:link
+php ${ARTISAN} db:seed --class CountriesSeederTable
```
Comment (djaiss) [inline docker/test-server.sh]: You also need to call `ActivityTypesTableSeeder` too.

Comment (djaiss) [inline docker/test-server.sh]: Otherwise it won't work when you add activities.

**Comment by kstrauser** [comment]
Thanks, @moosilauke18!

@djaiss The Docker Hub image is at https://hub.docker.com/r/monicahq/monicahq/ . It's my plan to give you the "monicahq" organization, though - I just registered it on your behalf. The branch probably isn't 100% ready to merge yet because I'm still making little tweaks. For instance, the seeding stuff. I feel like it's pretty close though!

File: docker/test-server.sh (hunk 2 above)
Comment (kstrauser) [inline docker/test-server.sh]: Fixed.

**Comment by adrum** [comment]
@kstrauser Do you think the image name on Docker Hub be monicahq/monica to match the Github structure? By the way, you've done a great job getting this going! I'm excited to give it a spin.

**Comment by adrum** [comment]
@kstrauser I just realized this project is called MonicaHQ (not Monica) from the README. I've just been surfing/monitoring the issues the past few days and haven't really spent much time on the README. Whoops.

**Comment by kstrauser** [comment]
@adrum That's alright, it was a surprise to me, too! The first several commits here named it "monica" until I saw a patch in the README to change some wording.

**Comment by kstrauser** [comment]
@djaiss Would you like me to squash these commits into a single patch? There are starting to be an awful lot of them. I don't have a preference.

**Comment by djaiss** [comment]
Yeah I'm still unsure of the project name to be honest: Monica or MonicaHQ, I hesitate.

@kstrauser for the squashing, I think it's ok as is, as anyway I'm squashing all commits before merging in order to keep a clean commit logs as much as possible.

Is it ready to be merged?

**Comment by adrum** [comment]
@djaiss My 2 cents is to go with just Monica. I think it's cleaner. The organization name makes sense as MonicaHQ. If you would like to discuss it further or want feedback on this, we can open a separate issue and discuss it there. I don't mean to clutter the Dockerize PR.

Btw, I used this branch from @kstrauser to get it up running in minutes. Took me a little bit to get the .env right. Mostly, the typical things from the README (APP_SECRET, MAIL_*, and APP_EMAIL_NEW_USERS_NOTIFICATION)

A few surprises: 

1. It would be nice to be able to change the DB credentials in .env during the first run via docker-compose. Perhaps by detecting if the storage volume is empty?

2. #153 differs from the .env settings `APP_EMAIL_NEW_USERS_NOTIFICATION`? It seems like these should be one settings, and by possibly adding APP_EMAIL_FROM_NAME?

**Comment by kstrauser** [comment]
I think @adrum makes a good point. I think "Monica, by MonicaHQ" has a nice sound.

Continuing the #153 discussion over there.

**Comment by djaiss** [comment]
We need to open a new issue for discussing the name I think. I want to address all the feminist remarks I've received and make sure that the name of the project doesn't offend anyone.

That being said, @kstrauser can I merge this?

**Comment by djaiss** [comment]
Also, if you want, I've added a CONTRIBUTORS file in the repository. If you wan tto appear, check this https://github.com/monicahq/monica/commit/47c5e74b3982a053684cb7c68c3bc6531df452bc

**Comment by kstrauser** [comment]
@djaiss It's rebased and (I believe) ready to merge.

**Comment by mottosso** [comment]
Hi all, great work with this, I was just about to dockerize it myself and couldn't have done a better job!

I'm however having trouble getting this up and running.

```bash
$ cp .env.example .env
$ docker-compose pull
$ docker-compose up
...
monicahq_1  | Unable to connect to MySQL; sleeping
monicahq_1  | Unable to connect to MySQL; sleeping
monicahq_1  | Unable to connect to MySQL; sleeping
```

I thought for a second that I needed a local install of MySQL, but I don't believe this to be the case.

On build, I'm also seeing this.

```bash
$ docker-compose build
...
Step 13/16 : RUN docker/install-composer.sh && ./composer.phar install
 ---> Running in 386a9ec54df4
/bin/sh: docker/install-composer.sh: not found
ERROR: Service 'monicahq' failed to build: The command '/bin/sh -c docker/install-composer.sh && ./composer.phar install' returned a non-zero code: 127
```

I'm building this on Windows 10 via the latest version of Docker Toolbox (which is itself running on VirtualBox).

Any ideas?

**Comment by mottosso** [comment]
I got things working, here's what I did.

1. Install MySQL on my local machine
2. Create and grant admin access to a new user
  ```bash
$ C:\Program Files\MySQL\MySQL Server 5.7\bin\mysql.exe
...
mysql> CREATE USER 'marcus'@'192.168.1.212' IDENTIFIED BY 'my_password';
mysql> GRANT ALL PRIVILEGES ON *.* TO 'marcus'@'localhost' WITH GRANT OPTION;
```
  - Where `192.168.1.212` is the IP of my local machine relative the running VirtualBox VM within which Docker is running.
3. Configure `.env` with the name and password of this new user
  ```bash
$ curl https://raw.githubusercontent.com/monicahq/monica/master/.env.example > .env
$ notepad .env
```
4. Run Docker
  ```bash
$ docker run -ti --rm --name monica --env-file .env -p 80:80 monicahq/monicahq
```
  - The `-ti` flag enables me to kill the instance via CTRL+C
  - The `--rm` flag automatically removes the image from disk once killed
  - The `--name` enables me to type `docker kill monica` as opposed to whatever name it is randomly assigned on creation.

**Comment by djaiss** [comment]
Lots of ❤️  for this PR. I'm mind blown.

**Comment by djaiss** [comment]
@kstrauser could you please send me the information for the Docker Hub? Thanks in advance and again thank you for your work.

**Comment by kstrauser** [comment]
@djaiss Kind of you to say, thanks! Register for a Docker Hub account and give me your username, and I'll add you as an owner to the organization.

**Comment by djaiss** [comment]
@kstrauser I'm `regisf` on Docker Hub. Thanks a lot!

**Comment by Phrohdoh** [comment]
@mottosso I am confused on step you noted.

> Where 192.168.1.212 is the IP of my local machine relative the running VirtualBox VM within which Docker is running.

Why are you running Docker inside of a VM? Is this absolutely necessary?
I was hoping to throw Docker onto an rpi or similar and run the project there.

**Comment by mottosso** [comment]
> Why are you running Docker inside of a VM? Is this absolutely necessary?

It's a Windows thing. For the longest time, Docker was limited to Linux and running it on Windows and OSX meant running it within a VM. This is their "Docker Toolbox" offering.

Nowadays there is a "Docker for Windows" which instead builds on top of Hyper-V, but the principle is the same in that you're essentially running a Linux VM within which the Docker Engine is running, unless you switch "mode" into running Windows containers instead.

> I was hoping to throw Docker onto an rpi or similar and develop/run the project there (in different container instances of course).

I think you can, in fact that's a good idea I might do the same thing; just happen to have a RPi running at home.

**Comment by kstrauser** [comment]
@djaiss I added you as an owner to the [monicahq organization](https://hub.docker.com/u/monicahq/dashboard/). Hey, look! 546 pulls!

If you register a [gravitar](https://en.gravatar.com) for MonicaHQ, we can set that as the project icon in Docker Hub.

**Comment by djaiss** [comment]
Thanks a lot! 546 pulls already, omg.

What do you think about using the Automated builds feature on Docker [doc](https://docs.docker.com/docker-cloud/builds/automated-build/#configure-automated-build-settings) by connecting GitHub to it? Would it work as is, or do I need to build the image myself?

**Comment by kstrauser** [comment]
That would be fantastic. I have a little Makefile at home that builds, tags, and pushes the upload, but I'd love it if it were automatic as part of the build process.

**Comment by djaiss** [comment]
@kstrauser so building the image is harder than simply connect Docker Hub to GitHub... I was dreaming :-D

**Comment by djaiss** [comment]
@kstrauser could you send me the Makefile to build images? I'll try to automate a daily build/upload. Thanks

**Comment by kstrauser** [comment]
@djaiss It's pretty tiny:

```Makefile
GIT_COMMIT=$$(git rev-parse --short HEAD)

docker_build:
	docker-compose build

docker_tag:
	docker tag monicahq/monicahq monicahq/monicahq:$(GIT_COMMIT)

docker_push:
	docker push monicahq/monicahq:$(GIT_COMMIT)
```

**Pull request merged**